driven/storage/adapter_v3.go

```diff
@@ -11,13 +11,154 @@ import (
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+// FindGroupsV3 finds groups with filter
+func (sa *Adapter) FindGroupsV3(clientID string, filter *model.GroupsFilter) ([]model.Group, error) {
+	groupIDs := []string{}
+	var userID *string
+	var memberships model.MembershipCollection
+
+	if filter != nil && filter.MemberUserID == nil && filter.MemberExternalID != nil {
+		var user model.User
+		err := sa.db.users.Find(bson.D{
+			{"client_id", clientID},
+			{"_id", filter.MemberExternalID},
+		}, &user, nil)
+		if err != nil {
+			userID = &user.ID
+		}
+	}
+	if userID == nil && filter != nil && filter.MemberUserID == nil && filter.MemberID != nil {
+		membership, _ := sa.FindGroupMembershipByID(clientID, *filter.MemberID)
+		if membership != nil {
+			userID = &membership.UserID
+			groupIDs = append(groupIDs, membership.GroupID)
+		}
+	}
+
+	if filter.MemberUserID != nil {
+		// find group memberships
+		memberships, err := sa.FindGroupMemberships(clientID, &model.MembershipFilter{})
```

Comment:
The filter being passed to this function does not contain anything so presumably it will return all group memberships? I assume we would want to pass `&model.MembershipFilter{UserID: filter.MemberUserID}`. Also, do we want to do this if we have already filtered by `MemberID` or `MemberExternalID`? It looks like you are trying to apply only one user related filter. It also seems like we should be setting `userID = filter.MemberUserID` here.

Reply:
Yes. This was one of few items that have been inconsistent. Thanks for those catches!